### ISSUE TITLE
Multiple VPA resources cause high load on the etcd

### ISSUE THREAD

**Issue by apolovov** [kubernetes/autoscaler#5421]
**Which component are you using?**:
vertical-pod-autoscaler

**What version of the component are you using?**:

Component version:
0.9.2
**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.14", GitCommit:"0f77da5bd4809927e15d1658fb4aa8f13ad890a5", GitTreeState:"clean", BuildDate:"2022-06-15T14:17:29Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.14", GitCommit:"0000000000000000000000000000000000000000", GitTreeState:"archive", BuildDate:"2022-09-01T10:43:00Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"linux/amd64"}</pre></details>

**What environment is this in?**:
bare metal

**What did you expect to happen?**:
vpa-recommender doesn't generate unwanted iopses in etcd.

**What happened instead?**:
I have vpa-recommender in cluster:
```
      - command:
        - /recommender
        - --address=:8942
        - --kube-api-burst=100
        - --kube-api-qps=50
        - --pod-recommendation-min-cpu-millicores=25
        - --pod-recommendation-min-memory-mb=50
        - --recommender-interval=30s
        - --stderrthreshold=0
        - --v=4
```
Also, i have 2k VPA-resources. With recommender-interval = 30s i have 300 iopses on master node. If i increase the interval to 5min, iopses decrease to 140:
![image](https://user-images.githubusercontent.com/18152319/212572062-b88858b8-fd46-4e9d-94d9-a12345078d7e.png)
Request rate to apiserver decreases too:
![image](https://user-images.githubusercontent.com/18152319/212572483-793b9a6e-0277-460b-983c-74bc6f7aab36.png)

It seems, that recommender cycle handler includes VPA patching and VPACheckpoint calculating. It is ok for me if VPA resources will lose some interactivity, but i need better VPACheckpoint resolution. So, i can't just increase the interval.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Comment by voelzmo** [comment]
Hey @apolovov, my understanding is that you're looking for a way to increase the frequency in which the recommender records VPACheckpoints and you would be okay if the VPA resource status would be updates less frequent (e.g. with the default 60s interval). Can you elaborate a bit what you are trying to achieve with decreasing the VPACheckpoint writing interval beyond the interval of calculating new recommendations?

As an aside: VPACheckpoints are updated based on PodMetrics received from the MetricsServer, which by [default also scrapes every 60s](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md#how-often-metrics-are-scraped), did you also configure a smaller scraping interval here?

**Comment by apolovov** [comment]
@voelzmo thank you for a reply!

> my understanding is that you're looking for a way to increase the frequency in which the recommender records VPACheckpoints and you would be okay if the VPA resource status would be updates less frequent

It is true, but i will refine a bit. I am trying to keep "VPACheckpoints frequency" equal to my prometheus scraping interval (30s), in the other hand it is okay if the VPA resource status would be updated less frequent — i believe, it affects the load and 60s interval doesn't decrease the iops significantly, but 5m does.

> Can you elaborate a bit what you are trying to achieve with decreasing the VPACheckpoint writing interval beyond the interval of calculating new recommendations?

I wish to have accurate statistics for fair resource allocation, at the same time, i don't really need high interactivity in recommendations.

> did you also configure a smaller scraping interval here?

Yes, but the default interval wouldn't help me much (I found out this experimentally).

**Comment by voelzmo** [comment]
Thanks, @apolovov, this helps me understand more about what you're doing there! I see you mentioning prometheus, does that mean you're specifying [`--storage=prometheus` ](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/recommender/main.go#L47)for the `vpa-recommender`?
If that is the case, your `vpa-recommender` doesn't use `VPACheckpoints`, but will use prometheus for retrieving historical data for the Pods.

**Comment by apolovov** [comment]
> does that mean you're specifying [--storage=prometheus ](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/recommender/main.go#L47)for the vpa-recommender?

No, i don't use the --storage=prometheus (it causes some weird problems when prometheus crashes) and my `vpa-recommender` does use `VPACheckpoints`.

I am sorry, i didn't mention it, but i use [prometheus metrics adapter](https://github.com/kubernetes-sigs/prometheus-adapter) to implement Metrics API. It is a proxy that converts the Metrics API requests to prometheus queries.

**Comment by voelzmo** [comment]
Ah, thanks, we're also using the prometheus-metrics-adapter, but only for custom metrics (everything in the `custom.metrics.k8s.io` group). I wasn't aware you can also use it as a metrics-server replacement (which serves the `metrics.k8s.io` group).

This setup still feels kind of weird in the sense that historical information is probably present in prometheus (depending on for how long you're keeping the data), and all that VPACheckpoints do is provide historical information in the case that vpa-recommender restarts. I'm not too familiar with the prometheus storage for the vpa-recommender and with using the prometheus-metrics-adapter to implement the Metrics API, so I don't understand the implications of prometheus crashing in these two settings very well.

To summarize my understand of what you're trying to achieve:
* You want to store historical information with a granularity of 30s, which means the vpa-recommender would need to do this in a 30s interval
  * retrieve PodMetrics
  * put them into its in-memory state
  * write the in-memory state in VPACheckpoints so the vpa-recommender can restore its in-memory state upon restart
 * You want to get new recommendations with a lower granularity, which means the vpa-recommender would need to do this in a 5m(?) interval
   * compute new recommendations based on the in-memory state (this could also be done every 30s, not sure how useful that would be)
   * update the status of all VPA resources with the new recommendations

Looking at this, I'm not sure how useful it is to retrieve PodMetrics with a much higher resolution than using it. What would be the drawbacks of also retrieving PodMetrics with the same 5m resolution? While vpa-recommender is up and running, you would miss very short-lived spikes in resources – but if you wouldn't get a new recommendation from seeing that spike, I'm not sure how useful having that data would be? Most people seem to try to get an accurate recommendation as quickly as possible, so they don't have to suffer from the negative effects of having too low resource requests (either OOMKills or latency increasing due to insufficient CPU). Prioritizing for the event of vpa-recommender restarting is an interesting choice I haven't seen so far.

Taking a step back and looking at how this issue started it seems that your initial goal is to reduce the IOPS on etcd. I think in a cluster with ~2k VPA resources having 300 IOPS seems quite reasonable to me? On busy clusters we see much higher mean IOPS, with peaks going >1k IOPS every few minutes.

**Comment by apolovov** [comment]
@voelzmo Thank you for comprehensive answer. You are right in summarizing my wishes. And i am agree with your arguments about usefulness (about detecting low resource limits and not-updating the vpa as soon as possible). And, to be honest, i cannot parry your arguments.

> I think in a cluster with ~2k VPA resources having 300 IOPS seems quite reasonable to me?

It is quite reasonable, but it hurts when i know there could be less and it doesn't feels fair that the half of iopses ate vpa. Also, we have hundreds of clusters and my pain is being multiplied :).

It seems, there isn't a solid reason to re-organize the vpa controller. Apparently, i will use the [flow-control feature](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/).

**Comment by voelzmo** [comment]
/kind support
/remove-kind bug
/close

Closing this, I think we've clarified the questions. Let me know if there's anything missing and/or re-open.
